Add support for ingress and automation for developers to test on minikube

alaypatel07 opened the pull request:
This depends on https://github.com/konveyor/crane-lib/pull/93

In order to test this execute the following steps:
1. install jq, minikube and kvm driver for minikube in the system
2. run `bash hack/minikube-clusters.sh`
3. wait for the ingress pods to be running with `kubectl get pods -n ingress-nginx --context=destination-cluster`
4. Create your workloads and identify 1 PVC to sync
5. Build crane with this PR and run `./crane  transfer-pvc --destination-context=destination-cluster --source-context=source-cluster --pvc-name mysql --pvc-namespace mysql`
6. You will see logs like this if the run is successful
```
2021/12/03 22:00:49 [114] Total bytes sent: 67.32K
Total bytes sent: 67.32K
2021/12/03 22:00:49 [114] Total bytes received: 211.77K
Total bytes received: 211.77K

2021/12/03 22:00:49 [114] sent 67.32K bytes  received 211.77K bytes  558.17K bytes/sec
sent 67.32K bytes  received 211.77K bytes  558.17K bytes/sec
2021/12/03 22:00:49 [114] total size is 8.60K  speedup is 0.03
total size is 8.60K  speedup is 0.03
```
7. clean up the environment by running `bash hack/delete-cluster.sh`

djzager reviewed (commented):
Just some comments on the scripts. Don't see anything too crazy in the transfer-pvc logic.

alaypatel07 commented on hack/minikube-clusters.sh:
```diff
@@ -0,0 +1,45 @@
+COREFILE=$(echo $COREFILE | sed s/'fallthrough\\n }\\n/& file \/etc\/coredns\/crane.db crane.dev\\n/')
+kubectl get cm -n kube-system coredns -ojson | jq ".data.Corefile = ${COREFILE}" | kubectl replace -f -
+
+kubectl patch cm -n kube-system coredns --type='json' -p='[{"op": "replace", "path": "/data/crane.db", "value": "; crane.dev test file\ncrane.dev.              IN      SOA     a.crane.dev. b.crane.dev. 2 604800 86400 2419200 604800\ncrane.dev.              IN      NS      a.crane.dev.\ncrane.dev.              IN      NS      b.crane.dev.\na.crane.dev.            IN      A       127.0.0.1\nb.crane ...
```
we need to edit the coredns networking for it to resolve the "crane.dev" domain that we are going to use in the ingress hostnames. Since these are our toy domains, I added a manual DNS entry to the database. We need the configmap with the following entry to make dns work

```
  crane.db: |
    ; example.org test file
    crane.dev.              IN      SOA     a.crane.dev. b.crane.dev. 2 604800 86400 2419200 604800
    crane.dev.              IN      NS      b.crane.dev.
    crane.dev.              IN      NS      a.crane.dev.
    a.crane.dev.            IN      A       127.0.0.1
    b.crane.dev.            IN      A       127.0.0.1     
    
    *.crane.dev.            IN      A       192.168.39.110
    ```

alaypatel07 commented on hack/minikube-clusters.sh:
```diff
@@ -0,0 +1,45 @@
+SOURCE_IP_RANGE=$(echo $SOURCE_IP | cut -d"." -f1-3).0/24 # subnet 11.1.1.0/24
+DEST_IP_RANGE=$(echo $DEST_IP | cut -d"." -f1-3).0/24 # subnet 11.1.1.0/24
+
+sudo iptables -I FORWARD 2 -p all -s $SOURCE_IP_RANGE -d $DEST_IP_RANGE -j ACCEPT
+sudo iptables -I FORWARD 3 -p all -s $DEST_IP_RANGE -d $SOURCE_IP_RANGE -j ACCEPT
```
Yes,  but IMO its non-persistent with respect to a reboot, but the minikube clusters will pause too, the scripts need to be re-run anyway after a reboot.

alaypatel07 commented:
Renamed the files, added the suggestions as fixups, one of them didnt work so reverted it back to what I had and responded to couple of questions PTAL @djzager

alaypatel07 commented on hack/minikube-clusters.sh:
```diff
@@ -0,0 +1,49 @@
+
+SOURCE_IP=$(minikube ip -p ${SRC_CLUSTER_NAME})
+DEST_IP=$(minikube ip -p ${DEST_CLUSTER_NAME})
+SOURCE_IP_RANGE="${SOURCE_IP#*.}.0/24" # subnet 11.1.1.0/24
```
discussed on slack, this is what we needed
```
SOURCE_IP_RANGE="${SOURCE_IP%.*}.0/24"
DEST_IP_RANGE="${DEST_IP%.*}.0/24"
```

Thanks @djzager for pointing it out. Updated as suggested

JaydipGabani commented on cmd/transfer-pvc/transfer-pvc.go:
```diff
@@ -312,6 +287,84 @@ func (t *TransferPVCOptions) run() error {
 		log.Fatal(err, "error following rsync client logs")
 	}
 
+	log.Println("followed the logs, garbage collecting created resources on both sourcen and destination")
```
s/sourcen/source

alaypatel07 replied:
fixed

djzager commented on cmd/transfer-pvc/transfer-pvc.go:
```diff
@@ -370,6 +427,80 @@ func followClientLogs(srcConfig *rest.Config, c client.Client, namespace string,
+
+func createAndWaitForRoute(pvc *corev1.PersistentVolumeClaim, destClient client.Client) endpoint.Endpoint {
+	// create a route for data transfer
+	// TODO: pass in subdomain instead of ""
```
Should the subdomain be accepted as an argument?

djzager commented on cmd/transfer-pvc/transfer-pvc.go:
```diff
@@ -370,6 +427,80 @@ func followClientLogs(srcConfig *rest.Config, c client.Client, namespace string,
 
+func createAndWaitForIngress(pvc *corev1.PersistentVolumeClaim, destClient client.Client) endpoint.Endpoint {
+	// create a route for data transfer
+	// TODO: add a config flag for subdomain
```
Same as below. Should the subdomain be accepted as an argument?
Taking care of this in this PR or follow-up?

alaypatel07 replied:
the PR that add the flag should do that IMO, do you mind if I create an issue for this to be taken care of as a follow up?

djzager replied:
works for me. add that as an issue.

djzager commented on hack/minikube-clusters-start.sh:
```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env bashset +x
+
+export KUBECONFIG=${PWD}/state-transfer.kubeconfig
```
Do we need this?

alaypatel07 replied:
this was just to make sure that the KUBECONFIG only has two contexts and we don't collide with a user-defined context that is the same name as ours.

djzager replied:
I think we should drop this for now. We can add notes about overriding KUBECONFIG to control where the file ends up , collisions, etc.

alaypatel07 replied:
dropped

djzager commented on hack/minikube-clusters-start.sh:
```diff
@@ -0,0 +1,49 @@
+
+kubectl patch deploy -n kube-system coredns --type='json' -p='[{"op": "add", "path": "/spec/template/spec/volumes/0/configMap/items/1", "value": {"key": "crane.db", "path": "crane.db"}}]'
+
+kubectl patch deploy --context=destination-cluster -n ingress-nginx ingress-nginx-controller --type='json' -p='[{"op": "add", "path": "/spec/template/spec/containers/0/args/12", "value": "--enable-ssl-passthrough"}]'
```
```suggestion
kubectl patch deploy --context=${DEST_CLUSTER_NAME} -n ingress-nginx ingress-nginx-controller --type='json' -p='[{"op": "add", "path": "/spec/template/spec/containers/0/args/12", "value": "--enable-ssl-passthrough"}]'
```

djzager commented on hack/minikube-clusters-start.sh:
```diff
@@ -0,0 +1,49 @@
+
+kubectl patch deploy -n kube-system coredns --type='json' -p='[{"op": "add", "path": "/spec/template/spec/volumes/0/configMap/items/1", "value": {"key": "crane.db", "path": "crane.db"}}]'
+
+kubectl patch deploy --context=destination-cluster -n ingress-nginx ingress-nginx-controller --type='json' -p='[{"op": "add", "path": "/spec/template/spec/containers/0/args/12", "value": "--enable-ssl-passthrough"}]'
+
+# force a rollout
+kubectl delete rs -n ingress-nginx --context=destination-cluster  -l app.kubernetes.io/component=controller,app.kubernetes.io/instance=ingress-nginx
```
```suggestion
kubectl patch deploy --context=${DEST_CLUSTER_NAME} -n ingress-nginx ingress-nginx-controller --type='json' -p='[{"op": "add", "path": "/spec/template/spec/containers/0/args/12", "value": "--enable-ssl-passthrough"}]'

# force a rollout
kubectl delete rs -n ingress-nginx --context=${DEST_CLUSTER_NAME}  -l app.kubernetes.io/component=controller,app.kubernetes.io/instance=ingress-nginx
```

alaypatel07 commented on hack/minikube-clusters-start.sh:
```diff
@@ -0,0 +1,49 @@
+
+export KUBECONFIG=${PWD}/state-transfer.kubeconfig
+
+SRC_CLUSTER_NAME=source-cluster
+DEST_CLUSTER_NAME=destination-cluster
```
+1 this seems reasonable shortcuts.

alaypatel07 commented on hack/minikube-clusters-start.sh:
```diff
@@ -0,0 +1,49 @@
+
+minikube update-context -p ${SRC_CLUSTER_NAME}
+
+# this hack does not work if the script is run twice
```
adding the following condition to the start of the script

if souce or destionation cluster exists, echo "run hack/minikube-clusters-delete.sh"; exit 0

Would that be acceptable?

JaydipGabani reviewed (approved):
VISACK! will test thoroughly
